### PR TITLE
feat: Search and Replace functionality

### PR DESCRIPTION
**Pull request by jmelnych**
This PR...

## Changes
Feat:
Add Search pane
Add Search functionality

Refactor:
Move out create Node functionality to a separate file
Remove color duplicate

_____
Must haves:
- [x] The styling of a search result when it's selected and highlighted should be improved to make the text easier to read.
- [x] When the user starts typing, I think we should show a skeleton or a loading spinner under the search input so the user knows something is happening.
- [x] When a result is selected, in the code editor we should select and scroll to the first match (this might be harder to do but we've done it already for Outgoing Links so maybe you could use that logic).

Nice haves:
- [x] The list of results should be more similar to how VSCode does it, each result becoming a dropdown
- [x] The listing the matched lines also highlighted in editor
- [x] Follow-up on the above, when clicking on a matched line, we should select and scroll to it in the editor.

## Screenshots

-

## Checklist

- [x] tested locally
- [x] added new dependencies
- [ ] updated the docs
- [ ] added a test

### CODE REVIEW

File: src/components/organisms/SearchPane/SearchPane.tsx [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+              <TreeItem
+                treeKey={String(event.key)}
+                title={event.title}
+                processingEntity={processingEntity}
+                setProcessingEntity={setProcessingEntity}
+                onDelete={onDelete}
+                onRename={onRename}
+                onExcludeFromProcessing={onExcludeFromProcessing}
+                onIncludeToProcessing={onIncludeToProcessing}
+                onCreateFolder={onCreateFolder}
```
Comment (jmelnych) [inline src/components/organisms/SearchPane/SearchPane.tsx:501]: Probably need to separate this functionality of all these onCreateFolder, onDelete, onPreview etc methods and reuse it both in FilterTree and SearchPane so not to duplicate all these methods in both components

File: src/components/organisms/PaneManager/PaneManagerLeftMenu.styled.tsx [hunk 2]
```diff
@@ -21,3 +23,7 @@ export const Container = styled.div<{isLeftActive: boolean}>`
   gap: 7px;
   background-color: ${props => (props.isLeftActive ? BackgroundColors.darkThemeBackground : PanelColors.toolBar)}};
 `;
+
+export const SearchIcon: React.FC = () => {
```
Comment (devcatalin) [inline src/components/organisms/PaneManager/PaneManagerLeftMenu.styled.tsx:27]: this could be added to the `@atoms/Icon` component

File: src/redux/thunks/updateFileEntry.ts [hunk 3]
```diff
@@ -39,6 +39,7 @@ export const updateFileEntry = createAsyncThunk(
         if (getFileStats(filePath)?.isDirectory() === false) {
           fs.writeFileSync(filePath, payload.content);
           fileEntry.timestamp = getFileTimestamp(filePath);
+          fileEntry.text = payload.content;
```
Comment (devcatalin) [inline src/redux/thunks/updateFileEntry.ts:42]: This might be outside the scope of the PR but we should probably rename `payload.content` to `payload.text` to avoid mistakes

File: src/components/organisms/SearchPane/SearchPane.tsx [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+  height: number;
+};
+
+type MatchParamProps = {
+  matchCase: boolean;
+  matchWholeWord: boolean;
+  regExp: boolean;
+};
+
+const SearchPane: React.FC<Props> = ({height}) => {
```
Comment (devcatalin) [inline src/components/organisms/SearchPane/SearchPane.tsx:60]: This file is quite large, I would recommend extracting logic into separate files as custom hooks, by grouping relevant useState + useEffect and other hooks together in custom ones.

**Review by devcatalin — CHANGES_REQUESTED** [review]

**Comment by devcatalin** [comment]
Great job with this one, I just tested it and it works nicely.

I have a couple of overall suggestions that can be included in this PR or in follow-up PRs:

Must haves:
- The styling of a search result when it's selected and highlighted should be improved to make the text easier to read.
- When the user starts typing, I think we should show a skeleton or a loading spinner under the search input so the user knows something is happening.
- When a result is selected, in the code editor we should select and scroll to the first match (this might be harder to do but we've done it already for Outgoing Links so maybe you could use that logic).

Nice haves:
- The list of results should be more similar to how VSCode does it, each result becoming a dropdown and then listing the matched lines of code to directly see what we've found without having to select and look through the file.
- Follow-up on the above, when clicking on a matched line, we should select and scroll to it in the editor.

File: src/components/organisms/PaneManager/PaneManagerLeftMenu.styled.tsx (hunk 2 above)
Comment (jmelnych) [reply src/components/organisms/PaneManager/PaneManagerLeftMenu.styled.tsx]: agree!

File: src/components/organisms/SearchPane/SearchPane.tsx (hunk 4 above)
Comment (jmelnych) [reply src/components/organisms/SearchPane/SearchPane.tsx:40]: did some refactoring for both SearchPane.tsx and FileTreePane.tsx, moved main chunks of logic to hooks/fileTreeHooks

File: src/redux/services/fileEntry.ts [hunk 5]
```diff
@@ -188,15 +190,24 @@ export function readFiles(
...
   } else {
     files.forEach(file => {
+      let text;
       const filePath = path.join(folder, file);
       const fileEntryPath = filePath.substring(rootFolder.length);
-      const fileEntry = createFileEntry({fileEntryPath, fileMap, helmChartId: helmChart?.id});
+
+      const isDir = getFileStats(filePath)?.isDirectory();
+
+      if (!isDir) {
```
Comment (devcatalin) [inline src/redux/services/fileEntry.ts:199]: maybe we should not read the content of files that are not "supported" or "included" ? based on `scanExcludes` and `fileIncludes`

File: src/redux/services/fileEntry.ts [hunk 6]
```diff
@@ -188,15 +190,24 @@ export function readFiles(
...
+        text = fs.readFileSync(path.join(filePath), 'utf8');
```
Comment (devcatalin) [inline src/redux/services/fileEntry.ts:200]: I think because we now read and save the text content of the file on the fileEntry, then in the `extractK8sResourcesFromFile` function from `services/fileEntry.ts` we should use this text directly instead of reading those files again. Because we don't want to read all files that contain resources two times instead of only once.

File: src/utils/getRegexp.ts [hunk 7]
```diff
@@ -0,0 +1,101 @@
+import {FileEntry} from '@models/fileentry';
+
+export type MatchParamProps = {
+  matchCase: boolean;
+  matchWholeWord: boolean;
+  regExp: boolean;
+};
+
+export function getRegexp(query: string, params: MatchParamProps): RegExp {
```
Comment (devcatalin) [inline src/utils/getRegexp.ts:9]: could you add a comment for this function with an example of input & output? something to show why this logic is needed

File: src/utils/getRegexp.ts [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+    queryRegExp = new RegExp(`\\b${query}\\b`, matchParams);
+  }
+  if (params.regExp) {
+    queryRegExp = new RegExp(query, matchParams);
+  }
+
+  return queryRegExp;
+}
+
+type Props = {
```
Comment (devcatalin) [inline src/utils/getRegexp.ts:30]: this looks similar to the MatchNode model, maybe it's not needed?
on the `decorateMatch` function below you could do:

```ts
function decorateMatch(text: string, query: string, fromIndex = 0): Omit<MatchNode, 'textWithHighlights'> {
```

File: src/utils/getRegexp.ts [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+      return {
+        ...node,
+        matches,
+        matchCount,
+        matchLines,
+      };
+    }
+
+    return null as unknown as FileEntry;
```
Comment (devcatalin) [inline src/utils/getRegexp.ts:97]: what does this do or why is it needed?

**Review by devcatalin — CHANGES_REQUESTED** [review]

File: src/components/organisms/SearchPane/RecentSearch.tsx [hunk 10]
```diff
@@ -0,0 +1,23 @@
+import electronStore from '@utils/electronStore';
+
+import * as S from './styled';
+
+type RecentSearchProps = {
+  handleClick: (query: string) => void;
+};
+
+const RecentSearch = ({handleClick}: RecentSearchProps) => {
+  const recentSearch: string[] = electronStore.get('appConfig.recentSearch')?.reverse() || [];
```
Comment (topliceanurazvan) [inline src/components/organisms/SearchPane/RecentSearch.tsx:10]: Should we maybe add a global variable for this, maybe it might be useful in other components?

File: src/hooks/fileTreeHooks/useCreate.tsx [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import {useAppDispatch} from '@redux/hooks';
+import {openCreateFolderModal, openNewResourceWizard} from '@redux/reducers/ui';
+
+export const useCreate = () => {
+  const dispatch = useAppDispatch();
+
+  const onCreateFolder = (absolutePath: string) => {
```
Comment (topliceanurazvan) [inline src/hooks/fileTreeHooks/useCreate.tsx:7]: Currently this doesn't exist on main anymore, was changed to `onCreateFileFolder`. We'll need to do some refactoring as we will have conflicts on merging

**Review by topliceanurazvan — CHANGES_REQUESTED** [review]

File: src/components/organisms/SearchPane/CreateFilteredNode.tsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        <S.EntryName className="file-entry-name">{item.name}</S.EntryName>
+        <S.MatchCount>{item.matchCount}</S.MatchCount>
+        <S.Path className="file-entry-path">{item.filePath}</S.Path>
+      </S.NodeTitleContainer>
+    </S.NodeContainer>
+  );
+
+  const StyledMatchLine = ({lineMatches}: MatchLineProps) => {
+    const wholeLine = lineMatches.reduce((acc: string, lm: MatchNode) => acc + lm.textWithHighlights, '');
+    return <S.MatchLine>{parse(wholeLine)}</S.MatchLine>;
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/components/organisms/SearchPane/CreateFilteredNode.tsx:30]: because we have <span> tag for css styling (yellow background in tree pane) to wrap around matched query;
React escapes all tags so it's either dangerously set HTML or a parser library

File: src/models/appstate.ts [hunk 13]
```diff
@@ -130,6 +130,8 @@ interface AppState {
   checkedResourceIds: string[];
   /** the currently selected path */
   selectedPath?: string;
+  /** the line number for the match in file */
+  lineNumber?: number;
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/models/appstate.ts]: yeap, already converted it to be a somewhat different object

File: src/hooks/fileTreeHooks/useProcessing.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+import {ExclamationCircleOutlined} from '@ant-design/icons';
+
+import {useAppDispatch, useAppSelector} from '@redux/hooks';
+import {updateProjectConfig} from '@redux/reducers/appConfig';
+import {scanExcludesSelector} from '@redux/selectors';
+
+export const useProcessing = (onOkHandler: () => void) => {
+  const scanExcludes = useAppSelector(scanExcludesSelector);
+  const configState = useAppSelector(state => state.config);
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/hooks/fileTreeHooks/useProcessing.tsx]: updated

File: src/hooks/fileTreeHooks/useHighlightNode.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+import {ROOT_FILE_ENTRY} from '@constants/constants';
+
+import {useAppDispatch, useAppSelector} from '@redux/hooks';
+import {setExpandedFolders} from '@redux/reducers/ui';
+
+import {TreeNode} from '@components/organisms/FileTreePane/types';
+
+import {uniqueArr} from '@utils/index';
+
+export const useHighlightNode = (tree: any, treeRef: any, expandedFolders: any) => {
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/hooks/fileTreeHooks/useHighlightNode.ts]: yes, forgot about providing types here! fixed

File: src/hooks/fileTreeHooks/useHighlightNode.ts [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+      treeRef?.current?.scrollTo({key: selectedPath});
+      return;
+    }
+
+    if (highlightNode) {
+      treeRef?.current?.scrollTo({key: highlightNode.key});
+    }
+  }, [tree]);
+
+  function highlightFilePath(filePath: string) {
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/hooks/fileTreeHooks/useHighlightNode.ts]: fixed

File: src/redux/services/fileEntry.ts (hunk 5 above)
Comment (jmelnych) [reply src/redux/services/fileEntry.ts]: updated

File: src/hooks/fileTreeHooks/useCreate.tsx (hunk 11 above)
Comment (jmelnych) [reply src/hooks/fileTreeHooks/useCreate.tsx]: rebased and fixed

File: src/utils/getRegexp.ts (hunk 9 above)
Comment (jmelnych) [reply src/utils/getRegexp.ts]: fixed

File: src/utils/getRegexp.ts (hunk 8 above)
Comment (jmelnych) [reply src/utils/getRegexp.ts]: fixed

File: src/redux/services/fileEntry.ts (hunk 6 above)
Comment (jmelnych) [reply src/redux/services/fileEntry.ts:202]: good point! reused the text

File: src/components/organisms/SearchPane/RecentSearch.tsx (hunk 10 above)
Comment (jmelnych) [reply src/components/organisms/SearchPane/RecentSearch.tsx]: good point, moved!

File: src/components/organisms/SearchPane/CreateFilteredNode.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (devcatalin) [reply src/components/organisms/SearchPane/CreateFilteredNode.tsx:30]: it's still not clear to me why we need this, could there be a solution with CSS only? I'm hesitant on using the `html-react-parser` or the dangerously set HTML

File: src/utils/replaceInFiles.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+
+export async function replaceInFiles(files: string[], query: RegExp, replaceQuery: string, dispatch: AppDispatch) {
+  const options = {
+    files,
+    from: query,
+    to: replaceQuery,
+  };
+  try {
+    const results = await replaceInFile(options);
+    console.log('Replacement results:', results);
```
Comment (topliceanurazvan) [inline src/utils/replaceInFiles.ts:16]: Lets use `log.info` from `loglevel`

File: src/utils/replaceInFiles.ts [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    const pathes = results.map(result => {
+      // const content = fs.readFileSync(result.file, 'utf8');
+      return {
+        relativePath: getRelativeFilePath(result.file),
+        absolutePath: result.file,
+      };
+    });
+    dispatch(updateFileEntries({pathes}));
+  } catch (error) {
+    console.error('Error occurred:', error);
```
Comment (topliceanurazvan) [inline src/utils/replaceInFiles.ts:26]: Same as above, use `log.error`

File: src/redux/thunks/updateFileEntry.ts [hunk 19]
```diff
@@ -108,6 +109,30 @@ export const updateFileEntry = createAsyncThunk(
...
+      try {
+        payload.pathes.forEach(ps => {
+          const fileEntry = mainState.fileMap[ps.relativePath];
+          const content = fs.readFileSync(ps.absolutePath, 'utf8');
+          fileEntry.text = content;
+          const newFileMap = {...mainState.fileMap, fileEntry};
+        });
+      } catch (e: any) {
+        const {message, stack} = e || {};
+        error = {message, stack};
```
Comment (topliceanurazvan) [inline src/redux/thunks/updateFileEntry.ts:137]: Should error be used anywhere, otherwise we can remove it

**Review by topliceanurazvan — CHANGES_REQUESTED** [review]

File: src/redux/thunks/updateFileEntry.ts [hunk 20]
```diff
@@ -108,6 +109,30 @@ export const updateFileEntry = createAsyncThunk(
...
+    const state: RootState = thunkAPI.getState();
+    let error: any;
+
+    const nextMainState: any = createNextState(state.main, mainState => {
+      try {
+        payload.pathes.forEach(ps => {
+          const fileEntry = mainState.fileMap[ps.relativePath];
+          const content = fs.readFileSync(ps.absolutePath, 'utf8');
+          fileEntry.text = content;
+          const newFileMap = {...mainState.fileMap, fileEntry};
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/redux/thunks/updateFileEntry.ts]: not really

File: src/components/organisms/SearchPane/CreateFilteredNode.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (jmelnych) [reply src/components/organisms/SearchPane/CreateFilteredNode.tsx:30]: @devcatalin, I need to style some specific part of the string which I wrap in a tag, in this case <em> tag. So far I don't see another way how I can embed it in the code without using html-react-parser or dangerouslySetHtml since the tag will appear as a plain text and not as a tag. Maybe I'm missing another solution for it... Any thoughts?

File: src/redux/thunks/updateFileEntry.ts (hunk 3 above)
Comment (jmelnych) [reply src/redux/thunks/updateFileEntry.ts]: refactored this part

**Review by topliceanurazvan — CHANGES_REQUESTED** [review]
Looking great 🙌🏻 Something small that I've noticed is that when you clear the search input, we still have a loading skeleton which I think should not appear as we are not searching until you type something
image

**Review by topliceanurazvan — APPROVED** [review]
Looking good from my side ! 👍🏻 

Great work !!!

File: package.json [hunk 21]
```diff
@@ -22,6 +22,7 @@
   },
   "private": true,
   "devDependencies": {
+    "@babel/plugin-proposal-private-property-in-object": "7.18.6",
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply package.json]: removed

File: package.json [hunk 22]
```diff
@@ -132,13 +134,15 @@
     "new-github-issue-url": "1.0.0",
     "node-fetch": "2.6.7",
     "node-machine-id": "1.1.12",
+    "npm": "8.13.2",
```
Comment (devcatalin) [inline package.json:137]: why is this needed?

File: package.json [hunk 23]
```diff
@@ -119,8 +120,9 @@
     "execa": "5.1.1",
     "fast-deep-equal": "3.1.3",
     "flat": "5.0.2",
-    "git-url-parse": "12.0.0",
     "framer-motion": "6.3.11",
+    "git-url-parse": "12.0.0",
+    "i": "0.3.7",
```
[the remark this answers is not in the document]
Comment (jmelnych) [reply package.json]: removed

File: package.json (hunk 22 above)
Comment (jmelnych) [reply package.json]: removed

**Review by devcatalin — APPROVED** [review]
Great work on this @jmelnych !
